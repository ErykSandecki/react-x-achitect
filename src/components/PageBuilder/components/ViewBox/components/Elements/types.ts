// types
import { MouseMode } from 'components/PageBuilder/enums';
import { TElement } from 'types';

export type TElementProps = {
  className: string;
  id: TElement['id'];
  mouseMode: MouseMode;
  parentId: TElement['parentId'];
  type: TElement['type'];
};
