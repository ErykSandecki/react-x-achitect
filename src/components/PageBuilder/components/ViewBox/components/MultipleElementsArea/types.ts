// types
import { TElement, TRectCoordinates } from 'types';

export type TElements = {
  elementsCordinates: Array<{
    coordinates: TRectCoordinates;
    id: TElement['id'];
  }>;
  outline: TRectCoordinates;
};
