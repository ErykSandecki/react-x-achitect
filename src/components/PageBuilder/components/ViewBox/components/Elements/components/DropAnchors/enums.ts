export enum DropAnchorsPosition {
  top = 'top',
  bottom = 'bottom',
  left = 'left',
  right = 'right',
}
