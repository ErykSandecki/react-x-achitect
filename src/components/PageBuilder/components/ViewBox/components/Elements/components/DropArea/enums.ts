export enum DropAreaPosition {
  top = 'top',
  bottom = 'bottom',
  left = 'left',
  right = 'right',
}
