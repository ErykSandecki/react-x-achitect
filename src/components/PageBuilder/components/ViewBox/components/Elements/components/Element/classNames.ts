export const className = 'Element';

export const classes = {
  className,
};

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      selected: `${className}--selected`,
    },
  },
};
