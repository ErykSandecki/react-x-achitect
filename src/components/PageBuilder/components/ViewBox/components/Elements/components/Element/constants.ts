export const OUTLINE = 2;
export const THROTTLE_WAIT = 20;
