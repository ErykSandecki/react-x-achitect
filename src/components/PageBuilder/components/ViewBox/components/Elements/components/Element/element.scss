@import '../../../../../../styles/z-indexes.scss';

.Element {
  $self: &;
  @include theme-selector('blue-1', outline, '&:hover', '1.5px solid');

  position: absolute;
  user-select: none;
  z-index: $--rx-zindex-element;

  &--selected {
    @include theme('blue-1', outline, '1.5px solid');
  }
}
