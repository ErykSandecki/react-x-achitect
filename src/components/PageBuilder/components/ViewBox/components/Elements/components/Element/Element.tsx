import { FC, memo, ReactNode, useRef } from 'react';
import { useSelector } from 'react-redux';

// components
import Corners from '../../../Corners/Corners';
import TransformArea from '../../../TransformArea/TransformArea';
import { Box } from 'shared';

// hooks
import { useElementEvents } from './hooks/useElementEvents';
import { useTheme } from 'hooks';

// others
import {
  className as classNameMoveableELement,
  classNames,
  classes,
} from './classNames';

// store
import {
  elementDynamicDataSelectorCreator,
  isSelectedElementSelectorCreator,
  multipleSelectedElementsSelector,
} from 'store/pageBuilder/selectors';

// styles
import styles from './element.scss';

// types
import { ElementType, TElement } from 'types';
import { MouseMode } from 'components/PageBuilder/enums';

// utils
import { getCornersPosition } from './utils/getCornersPosition';
import { isNumber } from 'lodash';

type TElementProps = {
  classes: typeof classes;
  children: (selected: boolean) => ReactNode;
  id: string;
  mouseMode: MouseMode;
  parentId: TElement['parentId'];
  type: ElementType;
};

const Element: FC<TElementProps> = ({
  classes,
  children,
  id,
  mouseMode,
  parentId,
  type,
}) => {
  const isSelected = useSelector(isSelectedElementSelectorCreator(id));
  const isMultiple = useSelector(multipleSelectedElementsSelector);
  const elementRef = useRef<HTMLDivElement>(null);
  const elementDynamicData = useSelector(elementDynamicDataSelectorCreator(id));
  const { coordinates } = elementDynamicData;
  const { backgroundColor, height, position, rotate, width } =
    elementDynamicData;
  const { x, y } = coordinates;
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);
  const rectCoordinates = getCornersPosition(height, width);
  const displayElements = !isMultiple && isSelected;
  const { isMoving, ...events } = useElementEvents(
    coordinates,
    elementRef,
    height,
    id,
    isMultiple,
    isSelected,
    mouseMode,
    parentId,
    position,
    type,
    width,
  );

  return (
    <Box
      classes={{
        className: cx(
          classes.className,
          classNamesWithTheme[classNameMoveableELement].name,
          [
            classNamesWithTheme[classNameMoveableELement].modificators.moving,
            isMoving,
          ],
          [
            classNamesWithTheme[classNameMoveableELement].modificators.selected,
            isSelected,
          ],
        ),
      }}
      id={id}
      ref={elementRef}
      style={{
        backgroundColor,
        height: isNumber(height) ? `${height}px` : height,
        left: `${x}px`,
        position,
        top: `${y}px`,
        transform: `rotate(${rotate}deg)`,
        width: isNumber(width) ? `${width}px` : width,
      }}
      {...events}
    >
      {children(isSelected)}
      {displayElements && <Corners rectCoordinates={rectCoordinates} />}
      {displayElements && (
        <TransformArea
          height={height}
          id={id}
          moseMode={mouseMode}
          x={x}
          y={y}
          width={width}
        />
      )}
    </Box>
  );
};

export default memo(Element);
