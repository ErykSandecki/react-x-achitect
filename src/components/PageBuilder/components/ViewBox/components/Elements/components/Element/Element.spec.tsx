import { Provider } from 'react-redux';
import { render } from '@testing-library/react';

// components
import Element from './Element';

// mocks
import {
  elementDynamicDataMock,
  elementStaticDataMock,
  pageBuilderStateMock,
  selectedElementMock,
} from 'test/mocks/reducer/pageBuilderMock';

// others
import { REDUCER_KEY as PAGE_BUILDER } from 'store/pageBuilder/actionsType';

// store
import { configureStore } from 'store/store';

// types
import { ElementType } from 'types';
import { MouseMode } from 'components/PageBuilder/enums';

const stateMock = {
  [PAGE_BUILDER]: {
    ...pageBuilderStateMock[PAGE_BUILDER],
    pages: {
      ...pageBuilderStateMock[PAGE_BUILDER].pages,
      ['0']: {
        ...pageBuilderStateMock[PAGE_BUILDER].pages['0'],
        elements: {
          ...pageBuilderStateMock[PAGE_BUILDER].pages['0'].elements,
          dynamicData: {
            [elementDynamicDataMock.id]: elementDynamicDataMock,
          },
          staticData: { [elementStaticDataMock.id]: elementStaticDataMock },
        },
      },
    },
  },
};

describe('Element snapshots', () => {
  it('should render Element', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <Element
          classes={{ className: 'className' }}
          id={selectedElementMock.id}
          mouseMode={MouseMode.default}
          parentId={selectedElementMock.parentId}
          type={ElementType.frame}
        >
          {() => <></>}
        </Element>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Element when is selected', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...stateMock[PAGE_BUILDER],
        pages: {
          ...stateMock[PAGE_BUILDER].pages,
          ['0']: {
            ...stateMock[PAGE_BUILDER].pages['0'],
            selectedElements: [selectedElementMock],
          },
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <Element
          classes={{ className: 'className' }}
          id={selectedElementMock.id}
          mouseMode={MouseMode.default}
          parentId={selectedElementMock.parentId}
          type={ElementType.frame}
        >
          {() => <></>}
        </Element>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Element when width & height is auto', () => {
    // mock
    const store = configureStore({
      ...stateMock,
      [PAGE_BUILDER]: {
        ...stateMock[PAGE_BUILDER],
        pages: {
          ...pageBuilderStateMock[PAGE_BUILDER].pages,
          ['0']: {
            ...pageBuilderStateMock[PAGE_BUILDER].pages['0'],
            elements: {
              dynamicData: {
                [elementDynamicDataMock.id]: {
                  ...elementDynamicDataMock,
                  height: 'auto',
                  width: 'auto',
                },
              },
              staticData: { [elementStaticDataMock.id]: elementStaticDataMock },
            },
          },
        },
      },
    });

    // before
    const { asFragment } = render(
      <Provider store={store}>
        <Element
          classes={{ className: 'className' }}
          id={selectedElementMock.id}
          mouseMode={MouseMode.default}
          parentId={selectedElementMock.parentId}
          type={ElementType.frame}
        >
          {() => <></>}
        </Element>
      </Provider>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
