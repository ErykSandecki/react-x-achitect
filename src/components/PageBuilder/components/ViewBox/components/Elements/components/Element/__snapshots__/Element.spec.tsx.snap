// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element snapshots should render Element 1`] = `
<DocumentFragment>
  <div
    class="css-q0izro className Element Element--light"
    data-status="false"
    data-test-box=""
    id="1"
    style="background-color: rgb(255, 255, 255); height: 100px; left: 0px; position: absolute; top: 0px; transform: rotate(0deg); width: 100px;"
  />
</DocumentFragment>
`;
